extensions/2.0/Vendor/STAIK_skeletons/README.md

```diff
@@ -0,0 +1,136 @@
+# STAIK_skeletons
+
+## Contributors
+
+* Marios Staikopoulos, <mailto:marios@staik.net>
+
+## Status
+
+Draft
+
+## Dependencies
+
+Written against the glTF 2.0 spec.
+
+## Overview
+
+This extension adds strict skeleton definitions to the glTF model. This allows renderers with skeletal based systems to avoid the ambiguity involved currently with skeleton derivation from skins.
+
+It also prevents the loss of joint data on export of specific skinned meshes:
+<p align="center">
+<img src="figures/skin-joints-tree.png" />
+</p>
+
+Unlike the base glTF 2.0 spec, the `joints` in the scene graph are no longer defined by a `skin`, they are now defined by the `skeleton`.
+Joints are unique to skeletons, and cannot be reused. While multiple skins can still use joints, the node ids in the skins `joints` array must all be nodes which belong to the skeletons `joints` array.
+
+This extension is meant to be *optional*. The current definitions allow importers which do not care about skeletons to fall back to their default skin-only implementation.
+
+## Nodes and Joints
+
+A node is a `joint` if the `skeleton` property is marked, and its value is the Id in the array of `skeletons` for which the joint is a member of.
+A `joint` must only contain transformation data. It cannot be a mesh, camera, or other entity.
+
+```json
+"nodes" : [
+    {
+        "name": "joint_0",
+        "extensions": {
+            "STAIK_skeletons": {
+                "skeleton": 0
+            }
+        }
+    }
+]
+```
+
+* **JSON schema**: [node.STAIK_skeletons.schema.json](schema/node.STAIK_skeletons.schema.json)
+
+## Skeletons
+
+Skeletons are composed of a tree of nodes, whose joints are nodes marked with the `skeleton` property.
+
+The subtree must:
+* Be a complete subtree (no non-joints lie in between joints of the skeleton)
+* Have a single root
+* The root must be joint
+* The root must be in the `joints` array.
```

Comment:
I'm not sure I understand why the extension needs to introduce a new `skeleton` concept to achieve these goals. Couldn't these same requirements be made mandatory for the joint subtree of a `skin`?

Reply:
The primary reason is that it would break almost all current implementations. Those points cause direct conflicts with the flexibility of the current specification, which many importers abuse.

Another reason is that it is perfectly OK for a mesh to only bind to a subset of joints, and not the entire skeleton. This allows smaller skins to be exported for larger skeletons, without exporting a bunch of identity IBM's.

Reply:
> The primary reason is that it would break almost all current implementations.

Our goal — I think —  is to restrict the existing specification for greater portability. On one hand, that is a breaking change by definition because it causes some assets that were previously valid to no longer be valid. On the other hand, clients that could load the existing specification will automatically be able to load a restricted subset of the existing specification. That second part is what we lose by introducing new concepts here rather than just adding restrictions. If implementing this extension is a no-op for existing clients, that is a significant advantage.

> Another reason is that it is perfectly OK for a mesh to only bind to a subset of joints, and not the entire skeleton.

This makes sense, but doesn't outweigh the concern above, to me anyway. There might be other missing features (shared skeletons, @bghgary?) that would justify the larger change.

> ...direct conflicts with the flexibility of the current specification, which many importers abuse.

I don't follow you here... do you mean *exporters* abuse it? My understanding was that we want exporters to write skeletons in a more tightly restricted way, so that importers will be simpler.

Reply:
Sorry about the delay, life decided to throw some curve-balls recently.

> I don't follow you here... do you mean exporters abuse it? My understanding was that we want exporters to write skeletons in a more tightly restricted way, so that importers will be simpler.

Sorry, yes that was a typo. It is exporters that are abusing the functionality. I do want to clarify that the *abuse* it self may or may not be intentional, it's just the ambiguity here that causes some exports to be so inconsistent.

For instance, different exporters exporting the same source file/data can produce different outputs.

As an example, take the same scene from Blender or Maya, whose joint trees are defined identically (albeit an armature in Blender vs just a joint-tree in Maya) and so are the skinned meshes.

The Blender exporter will export all of the armature joints as skin joints, even if they have zero-weightings.

The Maya2glTF exporter will only export the joints that have non zero-weights as skin joints. The rest will end up as regular nodes.

What happens is that we result in the same exact scene being exported in two different methods, and you cannot pin either exporter as being correct or incorrect, they are just behaving according to the spec allows.

> Our goal — I think —  is to restrict the existing specification for greater portability. On one hand, that is a breaking change by definition because it causes some assets that were previously valid to no longer be valid.

I agree with this, but from the feedback I received I was under the impression that such a change would have to get pushed to a major revision of glTF.

A counter point to this is that it can increase the perceived complexity of "skins", and potentially increase the difficulty for exporters to ensure that all of the skins are either subset of other skins, or define an entire new skin, as mentioned in my previous PR.

But as you said, shared skeletons/skins can also benefit from a separation of the definitions. If the skin joints are to used for say, indices in the skeleton, then you could re-use the same skin across multiple skeletons, as the skeleton would be a plug and play element.

> If implementing this extension is a no-op for existing clients, that is a significant advantage.

Yes, that was the intent of this extension. It should be perfectly viable for current node-based importers to throw away the existence of this plugin, and import as they always had.

However, it should aid tremendously in skeletal based importers.

I believe that the best solution for the spec would be to acknowledge the existence of skeletons as separate entities, and allow software that supports their existences to export them as such.

If you define the existence of skeletons, the hierarchy goes a bit like this:
`Nodes > Skeleton > Skins > Mesh`.

This still allows for importers with no limitations or need for skeletons to cut out the Skeleton component:
`Nodes > Skins > Mesh`

The existence of the skeleton can easily tell some importers whether or not they should even consider parsing skins from the glTF file.

Reply:
> shared skeletons, @bghgary?

FYI, #1285.